src/words/fast.jl

```diff
@@ -0,0 +1,193 @@
+# TODO: the input + idx combination should probably be replaced by a stream
+
+"""
+    TokenBuffer("foo bar")
+
+Turns a string into a readable stream, used for building tokenisers. Utility
+parser such as `spaces` and `number` read characters from the stream and
+into an array of tokens.
+
+Parsers return `true` or `false` to indicate whether they matched anything
+in the input stream. They can therefore be combined easily, e.g.
+
+    spaces(ts) || number(ts)
+
+either skips whitespace or parses a number token, if possible.
+"""
+mutable struct TokenBuffer
+  input::Vector{Char}
+  buffer::Vector{Char}
+  tokens::Vector{String}
+  idx::Int
+end
+
+TokenBuffer(input) = TokenBuffer(input, [], [], 1)
+
+TokenBuffer(input::AbstractString) = TokenBuffer(collect(input))
+
+Base.getindex(ts::TokenBuffer, i = ts.idx) = ts.input[i]
+isdone(ts::TokenBuffer) = ts.idx > length(ts.input)
+
+"""
+    flush!(::TokenBuffer, tokens...)
+
+TokenBuffer builds the current token as characters are read from the input. When
+the end of the current token is detected, call `flush!` to finish it and append
+it to the token stream. Optionally, give additional tokens to be added to the
+stream after the current one.
+"""
+function flush!(ts::TokenBuffer, s...)
+  if !isempty(ts.buffer)
+    push!(ts.tokens, String(ts.buffer))
+    empty!(ts.buffer)
+  end
+  push!(ts.tokens, s...)
+  return
+end
+
```

Comment:
Lets add a  `flush!(ts, chars::AbstractVector{Char}) = flush!(ts, String(chars))`

Then we can remove a lot of the calls to `flush!(ts, String(chars))`.